Add `gateBy` and `gate` to `FRP.Event.Time`

i-am-tom opened the pull request:
Previously, the functions only existed for `Behavior`. This PR adds them to `Event` as well.

@paf31 I don't know why we didn't do this before - can you shed some light? Happy to close this PR if I'm missing some `-behaviors` insight.

Commit:
Add `gateBy` and `gate` to `FRP.Event.Time`

Previously, the functions only existed for `Behavior`. This PR adds them
to `Event` as well.

paf31 commented:
Looks good, but why did you put these under `Time`?

i-am-tom commented:
Hah, late night copy-pasting 😳 I'll fix this at lunch in a couple hours.

Commit:
Fix comments from review

Put the functions in the wrong place. Oops. This commit moves the
`gate`-related functions from `FRP.Event.Time` to `FRP.Event.Class`, and
exports them.

paf31 commented on src/FRP/Event/Class.purs:
```diff
@@ -67,3 +69,21 @@ mapAccum f xs acc = filterMap snd
+gate :: forall a event. IsEvent event => event Boolean -> event a -> event a
+gate = gateBy const
+
+-- | Generalised form of `gateBy`, allowing for any predicate between
```
"Generalised form of `gate`"?

i-am-tom replied:
Oops, good catch!

paf31 commented on src/FRP/Event/Class.purs:
```diff
@@ -67,3 +69,21 @@ mapAccum f xs acc = filterMap snd
+  -> event b
+gateBy f s
+   = filtered
+ <<< sampleOn s
```
One issue here is that `b`s will be dropped until we see the first `a`. I think that needs fixing or documenting.

i-am-tom replied:
Yes, good point; which would you prefer? I'll take a look at the semantics, too!

paf31 replied:
I think I would prefer to not drop things initially, but it could even be indicated using a function argument.

Commit:
Final review changes, merging in blocking

As per the other PR, I've combined these two because there's some
interdependence. This implements the mentioned changes, and should now
be up-to-date with both PRs.

paf31 commented on src/FRP/Event/Time.purs:
```diff
@@ -15,3 +19,38 @@ foreign import animationFrame :: Event Unit
+debounce period = debounceWith (map { period, value: _ })
+
+-- | Provided an input event and transformation, block the input event for the
+-- | duration of the specified period on each output.
```
It might be nice to include a short example here, if there is one, and a simpler one above too.

Commit:
Replace "initial gating value" with `Maybe`

Previously, the function required an initial value. This wasn't great if
you didn't have such a value to hand, and has hence been removed in
favour of a `Maybe` in the predicate. The result is much neater :)

paf31 commented:
Thanks!
